app-backend/ingest/src/main/scala/com/azavea/rf/ingest/Ingest.scala

```diff
@@ -0,0 +1,179 @@
+import geotrellis.spark.io._
+import geotrellis.spark.io.file._
+import geotrellis.spark.io.index.ZCurveKeyIndexMethod
+import geotrellis.spark.io.s3._
+import geotrellis.spark.pyramid.Pyramid
+import geotrellis.vector.ProjectedExtent
+import geotrellis.raster._
+import geotrellis.raster.io.geotiff.MultibandGeoTiff
+import geotrellis.spark.tiling._
+import geotrellis.proj4.LatLng
+
+import org.apache.spark.rdd._
+import org.apache.spark._
+import spray.json._
+import DefaultJsonProtocol._
+import java.net.URI
+
+case class BandTile(band: Int, tile: Tile)
+
+object Ingest extends SparkJob {
+
+  case class Params(jobDefinition: URI = new URI(""), testRun: Boolean = false)
+  type RfLayerWriter = Writer[LayerId, RDD[(SpatialKey, MultibandTile)] with Metadata[TileLayerMetadata[SpatialKey]]]
+
+  /** Get a layerwriter and an attribute store for the catalog located at the provided URI */
+  def getRfLayerWriter(baseUri: URI): (RfLayerWriter, AttributeStore) = baseUri.getScheme match {
+    case "s3" | "s3a" | "s3n" =>
+      val (bucket, prefix) = S3.parse(baseUri)
+      val s3Writer = S3LayerWriter(bucket, prefix)
+      val writer = s3Writer.writer[SpatialKey, MultibandTile, TileLayerMetadata[SpatialKey]](ZCurveKeyIndexMethod)
+      (writer, s3Writer.attributeStore)
+    case "file" =>
+      val fileWriter = FileLayerWriter(baseUri.getPath)
+      val writer = fileWriter.writer[SpatialKey, MultibandTile, TileLayerMetadata[SpatialKey]](ZCurveKeyIndexMethod)
+      (writer, fileWriter.attributeStore)
+  }
+
+  /** Produce metadata for an IngestLayer */
+  def calculateTileLayerMetadata(layer: IngestLayer, scheme: LayoutScheme): (Int, TileLayerMetadata[SpatialKey]) = {
+    // We need to build TileLayerMetadata that we expect to start pyramid from
+    val overallExtent = layer.sources
+      .map(src => src.extent)
+      .reduce(_ combine _)
+      .reproject(LatLng, layer.output.crs)
+
+    // Infer the base level of the TMS pyramid based on overall extent and cellSize
+    val LayoutLevel(maxZoom, baseLayoutDefinition) =
+      scheme.levelFor(overallExtent, layer.output.cellSize)
+
+    maxZoom -> TileLayerMetadata(
+      cellType = layer.output.cellType,
+      layout = baseLayoutDefinition,
+      extent = overallExtent,
+      crs = layer.output.crs,
+      bounds = {
+        val GridBounds(colMin, rowMin, colMax, rowMax) =
+          baseLayoutDefinition.mapTransform(overallExtent)
+        KeyBounds(
+          SpatialKey(colMin, rowMin),
+          SpatialKey(colMax, rowMax)
+        )
+      }
+    )
+  }
+
+  /**  */
+  def multibandHistogram(rdd: RDD[(SpatialKey, MultibandTile)], numBuckets: Int): Vector[Histogram[Double]] =
+    rdd.map({ case (key, mbt) =>
+      mbt.bands.map { tile =>
+        tile.histogramDouble(numBuckets)
+      }
+    }).reduce({ (hs1, hs2) =>
+      hs1.zip(hs2).map { case (a, b) => a merge b }
+    })
+
+  /** We need to supress this warning because there's a perfectly safe `head` call being
```

Comment:
The multi-line comments need to be adjusted to follow our [styleguide](https://github.com/azavea/raster-foundry/blob/develop/docs/style/scala.md#comments)